### PR TITLE
Make Trash Usable - UI

### LINKED ISSUE #41665: FE: Correctly display archived/trashed collections, dashboards, and cards
```[tasklist]
### Tasks
```

### PR DESCRIPTION
**Pull request by sloansparger**
Closes #41665

### Description

This PR adds the new Trash experience in favor of the old Archive experience. An overview of the changes can be viewed in this [Figma design](https://www.figma.com/file/HRopvhNDlPOsBUQR8UMuOh/Make-Archive-actually-usable?type=design&node-id=226-9166&mode=design&t=wVJdy0W1gbDvRVKl-4). 

This PR does not remove the old archive UI, but it is currently unreachable except through direct route access. That part of the UI will get ripped out in a follow up PR.

### Demo
Screenshot 2024-05-02 at 12 09 05 PM

Screenshot 2024-05-02 at 12 09 39 PM

Screenshot 2024-05-02 at 12 11 29 PM

Screenshot 2024-05-02 at 12 10 42 PM
afe-ada2-ead24dfa156d">

Screenshot 2024-05-02 at 12 10 53 PM

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (rafpaf) [inline frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:36]:
```suggestion
            {t`This ${entity} is in the trash.`}
```

Comment (rafpaf) [reply frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:36]: For i18n I'd love if this provided each of the four possible strings to `t`: "This collection is in the trash.", etc. That way it can be translated properly into Romance languages among others

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (rafpaf) [inline frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:63]: If there's a chance that the icon might shrink when it container shrinks, maybe use `FixedSizeIcon`.

File: frontend/src/metabase/common/components/EntityPicker/components/EntityPickerModal/ButtonBar.tsx
Comment (npfitz) [inline frontend/src/metabase/common/components/EntityPicker/components/EntityPickerModal/ButtonBar.tsx:62]: whoops

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (rafpaf) [inline frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:45]: <3

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (rafpaf) [inline frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:48]: Would it be easy to do:

```
{"dashboard": t`This dashboard is in the trash.`, "collection": t`This collection is in the trash.`, etc.}[entityType]
```

Then the translations will be accurate

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (rafpaf) [inline frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:64]: nit: you could use Flex spacing here rather than marginInlineEnd

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (rafpaf) [inline frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:80]: same as above

File: frontend/src/metabase/collections/components/ActionMenu/ActionMenu.tsx
Comment (rafpaf) [inline frontend/src/metabase/collections/components/ActionMenu/ActionMenu.tsx:150]: maybe "This item has been..." for friendliness?

File: frontend/src/metabase/collections/components/BulkActions.tsx
Comment (rafpaf) [inline frontend/src/metabase/collections/components/BulkActions.tsx:133]: Would it be straightforward to drop in some ngettext pluralization?

File: frontend/src/metabase/collections/components/BulkActions.tsx
Comment (rafpaf) [inline frontend/src/metabase/collections/components/BulkActions.tsx:174]: the silently passed event argument catches my eye but probably no biggie

File: frontend/src/metabase/collections/components/BulkActions.tsx
Comment (rafpaf) [inline frontend/src/metabase/collections/components/BulkActions.tsx:241]: like above, maybe nggettext pluralization would work well here

File: frontend/src/metabase/components/EntityItem/EntityItem.tsx
Comment (rafpaf) [inline frontend/src/metabase/components/EntityItem/EntityItem.tsx:160]:
```suggestion
        title: isBookmarked ? t`Remove from bookmarks` : c('Verb').t`Bookmark`,
```

File: frontend/src/metabase/components/EntityItem/EntityItem.tsx
Comment (rafpaf) [inline frontend/src/metabase/components/EntityItem/EntityItem.tsx:198]:
```suggestion
        title: c('Verb').t`Duplicate`,
```

File: frontend/src/metabase/components/ItemsTable/BaseItemsTable.tsx
Comment (rafpaf) [inline frontend/src/metabase/components/ItemsTable/BaseItemsTable.tsx:207]:
```suggestion
              {isTrashed ? c("Precedes the name of a user").t`Deleted by` : t`Last edited by`}
```

File: frontend/src/metabase/components/ItemsTable/BaseItemsTable.tsx
Comment (rafpaf) [inline frontend/src/metabase/components/ItemsTable/BaseItemsTable.tsx:214]:
```suggestion
              {isTrashed ? c("Precedes the date and time of deletion").t`Deleted at` : t`Last edited at`}
```

File: frontend/src/metabase/dashboard/actions/trash.ts
Comment (rafpaf) [inline frontend/src/metabase/dashboard/actions/trash.ts:28]: Word-for-word translations, oh dear. I'll leave it up to you whether this can be remedied

File: frontend/src/metabase/dashboard/actions/trash.ts
Comment (rafpaf) [inline frontend/src/metabase/dashboard/actions/trash.ts:34]:
```suggestion
        // @ts-expect-error rtk typings are wrong - function expects an argument but TS errors saying it shouldn't have one
```

Should the string that identifies the specific error be included here?

File: frontend/src/metabase/entities/collections/collections.ts
Comment (rafpaf) [inline frontend/src/metabase/entities/collections/collections.ts:109]: localize the strings? Not sure if that is needed here

File: frontend/src/metabase/entities/dashboards.js
Comment (rafpaf) [inline frontend/src/metabase/entities/dashboards.js:87]: same question as above about localizing

**Review by rafpaf — COMMENTED** [review]
first pass

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (sloansparger) [reply frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx]: fixed, also flag to design that this is gross looking anyways. i should have caught this.

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (sloansparger) [reply frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:50]:
```
c("{0} is the entity in the trash, e.g. collection, dashboard, etc.").t`This ${entityType} is in the trash.`
```
How does this sound? I think we want to expand this to snippets, actions, etc. in the future. I'd prefer to keep the code where I don't have to come back and change it.

Comment (rafpaf) [reply frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:50]: My general view is that we should support i18n to a "good enough" standard, but you as the PR author basically get to decide what "good enough" means. So this is totally up to you.

My two cents, which you are free to take or leave: I think it's not too much of a headache to provide excellent support to the translators here, and provide a fallback for new entityTypes added later. That way, if you remember to come back and spell out the msgid for "snippet", "action", etc., great, but in the meantime those can use "this {0} is in the trash". 

Why I care: A sentence with a grammatical gender mismatch, like "Ce collection est à la poubelle," sounds pretty dumb, like "This collection are in the trash". I think we should avoid generating it if we can do so without twisting the code into terrible knots

File: frontend/src/metabase/dashboard/actions/trash.ts
Comment (sloansparger) [reply frontend/src/metabase/dashboard/actions/trash.ts]: Well, we had a requirements change, so I was able to fix this
```ts
export function getTrashUndoMessage(name: string, archived: boolean): string {
  return archived
    ? c(
        "{0} is the name of the entity being trashed, e.g. 'My Awesome Dashboard'",
      ).t`${name} has been moved to the trash.`
    : c(
        "{0} is the name of the entity being restored from the trashed, e.g. 'My Awesome Dashboard'",
      ).t`${name} has been restored.`;
}
```

Comment (rafpaf) [reply frontend/src/metabase/dashboard/actions/trash.ts]: re-read dat guy: "restored from the trashed"

Comment (rafpaf) [reply frontend/src/metabase/dashboard/actions/trash.ts]: btw the single quotes in the argument to `c` might trip up the verify-i18n step of CI

File: frontend/src/metabase/entities/collections/collections.ts
Comment (sloansparger) [reply frontend/src/metabase/entities/collections/collections.ts]: resolved using the function i inlined a few comments before this one

File: frontend/src/metabase/entities/dashboards.js
Comment (sloansparger) [reply frontend/src/metabase/entities/dashboards.js]: resolved using the function i inlined a few comments before this one

File: frontend/src/metabase/dashboard/components/DashboardInfoSidebar/DashboardInfoSidebar.tsx
Comment (iethree) [inline frontend/src/metabase/dashboard/components/DashboardInfoSidebar/DashboardInfoSidebar.tsx:73]: I think caching should also be gated behind can_write?

File: frontend/src/metabase/ui/components/icons/Icon/icons/trash_filled.svg
Comment (iethree) [inline frontend/src/metabase/ui/components/icons/Icon/icons/trash_filled.svg:1]: the filled version looks like a large pot of soup

**Review by iethree — DISMISSED** [review]
🏆  This is really great work. Didn't see anything that needs to block, but had a good pile of suggestions. 


Please wait for @metabase/querying-components to take a look or say they don't care.

File: frontend/src/metabase-types/api/collection.ts
Comment (iethree) [reply frontend/src/metabase-types/api/collection.ts:115]: oh wait, this is a blocker.

**Review by iethree — CHANGES_REQUESTED** [review]
looks good, the one blocker is having [non-api types in the api types file](https://github.com/metabase/metabase/pull/41666#discussion_r1589538101)

**Review by ranquild — APPROVED** [review]
metabase-lib/v1 changes LGTM

File: frontend/src/metabase-lib/v1/Question.ts
Comment (sloansparger) [reply frontend/src/metabase-lib/v1/Question.ts:266]: looks like we got approval 👍

File: frontend/src/metabase/collections/components/ActionMenu/ActionMenu.tsx
Comment (sloansparger) [reply frontend/src/metabase/collections/components/ActionMenu/ActionMenu.tsx:155]: Hmmm... I think the be I know how to do is
```
const Entity = entityForObject(item);
dispatch(Entity.actions.delete(item));
```
Would that be fine? (This is actually what the Search action here is doing... I've just yanked the code here from the existing Archive solution.)

Comment (iethree) [reply frontend/src/metabase/collections/components/ActionMenu/ActionMenu.tsx:155]: I think i like that better since it doesn't make me thing we're calling `DELETE /api/search/123` but up to you, whatever you think is clearer.

Comment (sloansparger) [reply frontend/src/metabase/collections/components/ActionMenu/ActionMenu.tsx:155]: I want to get this all ported over to RTKQuery anyways, so I'll run with the solution for now.

File: frontend/src/metabase/collections/components/BulkActions.tsx
Comment (sloansparger) [reply frontend/src/metabase/collections/components/BulkActions.tsx:1]: the link seems to be broken, but i have split the component up between archived / unarchived entities

File: frontend/src/metabase/api/collection.ts
Comment (sloansparger) [reply frontend/src/metabase/api/collection.ts:98]: I can think of a few options:
- blast the collections cache on each delete
- tag all collections with tags for their parent collections
- manipulate the rtkq cache directly

the first option solve delete, but we obviously missed this around update, when you could reparent a collection when moving and thus the effective location would change for all the subcollections in a collection

the second option seems is a memory intensive and cpu expensive solution as you generative the current depth's # of tags. my guts says for large instances we'd cause pauses while computing this when loading all collections, but i could be wrong.

the third option feels like a recipe for disaster imo as manual cache updates are keyed by endpoints, so as we add new endpoints we need to make sure all others are correctly updated. could be a common function, but still feels cumbersome in a way that the second approach wouldn't be.

File: frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx
Comment (sloansparger) [reply frontend/src/metabase/archive/components/ArchivedEntityBanner/ArchivedEntityBanner.tsx:41]: I don't remember the reason anymore, but Box is working. Thanks for noticing this!

File: frontend/src/metabase/components/ArchiveCollectionModal/ArchiveCollectionModal.jsx
Comment (sloansparger) [reply frontend/src/metabase/components/ArchiveCollectionModal/ArchiveCollectionModal.jsx:1]: Can confirm this is pre-existing. It actually works the same way for the move and permissions options. I'll make note of this so we come back and make this a better experience all around.

File: frontend/src/metabase/query_builder/actions/core/updateQuestion.ts
Comment (sloansparger) [reply frontend/src/metabase/query_builder/actions/core/updateQuestion.ts:256]: ah, it was only happening on questions but not dashboards. good catch!

File: frontend/src/metabase/components/ConfirmDeleteModal.tsx
Comment (iethree) [inline frontend/src/metabase/components/ConfirmDeleteModal.tsx:1]: ❤️ can you maybe publicize this in #engineering-frontend so we can stop the exponential growth of delete confirm modals in the codebase?

**Review by iethree — APPROVED** [review]
🥳

**Pull request merged**
